### PR TITLE
Support for TensorFunctionSpaces

### PR DESCRIPTION
**Pull request by ctjacobs**
* Add support for `TensorFunctionSpace`s, tensor `Functions` and tensor `Expressions`.
* Add support for tensor interpolation and projection.
* Add some tensor-related tests.

Addresses issue #348.

### CODE REVIEW

File: firedrake/functionspace.py [hunk 1]
```diff
@@ -599,6 +599,45 @@ def __getitem__(self, i):
...
+        shape = shape or (mesh.ufl_cell().geometric_dimension(), )*2
+
+        if isinstance(mesh, mesh_t.ExtrudedMesh):
+            raise NotImplementedError("TFS on extruded meshes not implemented")
+        else:
+            element = ufl.TensorElement(family, domain=mesh.ufl_domain(),
+                                        degree=degree, shape=shape,
+                                        symmetry=symmetry)
+        super(TensorFunctionSpace, self).__init__(mesh, element, name,
+                                                  dim=shape, rank=2)
```
Comment (wence-) [inline firedrake/functionspace.py]: Should rank always be 2, or should it be `len(shape)`.  I note a (somewhat cryptic) comment in `FunctionSpaceBase` says "this is the function space rank, the value rank may be different".  Do all tensor spaces have rank 2, irrespective of their shape?  Or does a TFS(...., shape=(1, 2, 3)) have rank 3?

File: tests/regression/test_function.py [hunk 2]
```diff
@@ -25,6 +39,40 @@ def test_firedrake_function(V):
...
+
+    g = Function(f)
+    assert np.allclose(g.dat.data_ro, vals)
+
+    # Check that g is indeed a deep copy
+    fvals = vals + 10
+    f.interpolate(Expression(fvals.astype("string")))
+
+    assert np.allclose(f.dat.data_ro, fvals)
+    assert np.allclose(g.dat.data_ro, vals)
```
Comment (wence-) [inline tests/regression/test_function.py]: I tidied these tests a little so that the expressions weren't repeated so much.

**Comment by wence-** [comment]
This basically looks good I think.  My only concern is raised above the "rank" comment.  Anyone else?

File: tests/regression/test_expressions.py [hunk 3]
```diff
@@ -33,6 +33,12 @@ def vfs(request, vcg1, cg1vcg1):
             'cg1vcg1[1]': cg1vcg1[1]}[request.param]
 
 
+@pytest.fixture(scope='module', params=['tcg1'])
+def tfs(request, tcg1):
+    """A parametrized fixture for tensor function spaces."""
+    return {'tcg1': tcg1}[request.param]
```
Comment (wence-) [inline tests/regression/test_expressions.py]: FWIW, given that there's only one tensor space, you could have written:
```
@pytest.fixture(scope='module')
def tfs(tcg1):
    return tcg1
```

But it's not a show-stopper :p

File: firedrake/functionspace.py (hunk 1 above)
Comment (ctjacobs) [inline firedrake/functionspace.py]: This should be `len(shape)` I think.

**Comment by wence-** [comment]
OK, I'm happy, will merge (travis failure is unrelated).

**Pull request merged**